internal/k8s/controllers/layer2_status_controller.go

```diff
@@ -99,32 +112,44 @@ func (r *Layer2StatusReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	}
 
 	level.Debug(r.Logger).Log("controller", "Layer2StatusReconciler", "updated state", dumpResource(state))
-
 	return ctrl.Result{}, nil
 }
 
 func (r *Layer2StatusReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	p := predicate.NewPredicateFuncs(func(object client.Object) bool {
 		if s, ok := object.(*v1beta1.ServiceL2Status); ok {
-			return strings.HasSuffix(s.Name, r.NodeName)
+			// only objects with complete labels that can illustrate which service it is related to
+			// can trigger the reconciler
+			label := s.GetLabels()
+			if label == nil {
+				return false
+			}
+			if _, ok = label[LabelAnnounceNode]; !ok {
+				return false
+			}
+			if _, ok = label[LabelServiceName]; !ok {
+				return false
+			}
+			if _, ok = label[LabelServiceNamespace]; !ok {
+				return false
+			}
 		}
 		return true
 	})
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&v1beta1.ServiceL2Status{}).
-		WatchesRawSource(&source.Channel{Source: r.ReconcileChan}, handler.EnqueueRequestsFromMapFunc(
+		Named("servicel2status").
+		// for crs, we build meta from cr label which indicate the service information
+		Watches(&v1beta1.ServiceL2Status{}, handler.EnqueueRequestsFromMapFunc(
 			func(ctx context.Context, object client.Object) []reconcile.Request {
-				evt, ok := object.(*l2StatusEvent)
-				if !ok {
-					level.Error(r.Logger).Log("controller", "Layer2StatusReconciler", "received an object that is not a l2StatusEvent from channel", "object", object)
-					return []reconcile.Request{}
-				}
-				level.Debug(r.Logger).Log("controller", "Layer2StatusReconciler", "enqueueing", "object", evt)
-				// unify req.Name to svcName-nodeName
+				level.Debug(r.Logger).Log("controller", "Layer2StatusReconciler", "enqueueing", "object", object)
+				label := object.GetLabels()
```

Comment:
You mean we don't have to filter by the labels'(svcName, svcNamespace) existence?
@oribon

Reply:
I think this is still the same since last time

Reply:
> I think this is still the same since last time

Oh, I misunderstood you and thought you were talking about the predicate part.

This part of code(in `Watches(&v1beta1.ServiceL2Status{}, ...)`) is not designed for validation, but for building reconciler request from cr label insdead of cr metadata, which is what we discussed before at the very beginning of this pr to solve some tricky problem.

So I think we should stick to the current `Watches(&v1beta1.ServiceL2Status{}, ...)` part and revert some of my previous commits which deleted some validation of label in predicate part due to my misunderstanding.

wdyt @fedepaol

Reply:
sorry about the earlier comment, wdyt about having the predicate act on the status fields after we make them immutable? cc @fedepaol